include/envoy/network/socket.h

```diff
@@ -0,0 +1,179 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/common/platform.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/core/v3/base.pb.h"
+#include "envoy/network/address.h"
+#include "envoy/network/io_handle.h"
+
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Network {
+
+// SocketOptionName is an optional value that captures the setsockopt(2)
+// arguments. The idea here is that if a socket option is not supported
+// on a platform, we can make this the empty value, which allows us to
+// avoid #ifdef proliferation.
+struct SocketOptionName {
+  SocketOptionName() = default;
+  SocketOptionName(const SocketOptionName&) = default;
+  SocketOptionName(int level, int option, const std::string& name)
+      : value_(std::make_tuple(level, option, name)) {}
+
+  int level() const { return std::get<0>(value_.value()); }
+  int option() const { return std::get<1>(value_.value()); }
+  const std::string& name() const { return std::get<2>(value_.value()); }
+
+  bool has_value() const { return value_.has_value(); }
```

Comment:
sorry, clang-tidy doesn't like this due to Envoy naming convention. This was in the codebase already but clang-tidy offers new levels of scrutiny when code moves to a new file.

You should probably just call this hasValue(), and update call-sites.

Reply:
+1 if you could pay it forward we would appreciate it.

Reply:
No worries! Updated patch!